---
title: "Class11: Genomic Informatics"
author: "Siyu Xie (PID: A16438448)"
format: pdf
toc: true
---

# Section 1. Proportion of G|G in MXL population


Here we read this CSV file

```{r}
MXL <- read.csv("373531-SampleGenotypes-Homo_sapiens_Variation_Sample_rs8067378.csv")
head(MXL)
```

```{r}
MXL$Genotype..forward.strand.
table(MXL$Genotype..forward.strand.)
```

```{r}
table(MXL$Genotype..forward.strand.)/nrow(MXL)*100
```

Now lets look at a different population.

```{r}
GBR <- read.csv("373522-SampleGenotypes-Homo_sapiens_Variation_Sample_rs8067378.csv")
```

```{r}
head(GBR)
```

Find proportion of G\|G of GBR

```{r}
round(table(GBR$Genotype..forward.strand.)/nrow(GBR)*100,2)
```

This variant associated with childhood asthma is more frequent in the
GBR than in MXL population.

Lets now dig into this further.

## Section 4 Population Analysis

One sample is obviously not enough to know what is happening in a
population. You are interested in assessing genetic differences on a
population scale.

So, you processed about ~230 samples and did the normalization on a
genome level. Now, you want to find whether there is any association of
the 4 asthma-associated SNPs (rs8067378...) on ORMDL3 expression.

> Q13.Read this file into R and determine the sample size for each
> genotype and their corresponding median expression levels for each of
> these genotypes

```{r}
expr <- read.table("rs8067378_ENSG00000172057.6.txt")
head(expr)
```

```{r}
library(dplyr)
expr %>%
  group_by(geno) %>%
  summarize(
    sample_size = n (),
    median_expression = median (exp, na.rm=TRUE)
  )
```

> Q14. Generate a boxplot with a box per genotype, what could you infer
> from the relative expression value between A/A and G/G displayed in
> this plot? Does the SNP effect the expression of ORMDL3?

A: the median level of expression for A|A is significantly higher than
G|G, meaning the SNP does affect the expression of ORMDL3.

```{r}
library(ggplot2)
```

```{r}
ggplot(expr) +
  aes(x = geno, y = exp, fill = geno)+
  geom_boxplot(notch = TRUE)

```
